README.md
# Kernel32Lib: kernel32lib

Win32 API wrapper for Ruby which uses Fiddle and Fiddle::Importer

## Installation and usage

install:
```
gem install kernel32lib
```
usage:
```ruby
  require 'kernel32lib/files'
  require 'kernel32lib/consoles'

  # call Kernel32Lib.function_name(function_params (optional)...)
  # e.g.
  Kernel32Lib.get_file_information_by_handle(__FILE__)
  # or
  Kernel32Lib.get_console_output_cp
```
see more detailed example in ./fixtures/fixture_kernel32lib.rb

## Troubleshooting

Visit to [kernel32lib homepage](https://github.com/dim11981/kernel32lib)


CHANGELOG.md
## 0.0.3 (Jul 9, 2016)

Features:
    - add WinAPI error codes translation (0-499)
    - add support file path in UNICODE

Bugfixes:
    - check file path encoding

## 0.0.2 (Jun 2, 2016)
Features:
- add WinAPI Consoles functions:
* GetStdHandle
* SetConsoleCursorPosition
* GetConsoleScreenBufferInfo
* SetConsoleScreenBufferSize
* SetConsoleWindowInfo
* SetConsoleTextAttribute
* GetConsoleOutputCP
* SetConsoleOutputCP
Bugfixes:
- check get_file_information_by_handle

## 0.0.1 (Jun 1, 2016)
Features:
- add WinAPI Files functions:
* CreateFile
* CloseHandle
* GetFileInformationByHandle
* FileTimeToSystemTime
* GetFileTime
- add module Fixture (fixture for test)
Bugfixes:
- initial release


test/test_kernel32lib.rb
# encoding: utf-8

require(File.expand_path('../../fixtures/fixture_kernel32lib',__FILE__))

# begin test
Fixture.show
# end test

win_api/files.rb
# encoding: utf-8

require(File.expand_path('../../lib/kernel32lib.rb',__FILE__))

# Kernel32Lib module
# Files
module Kernel32Lib
  extern 'void* CreateFile(char*,unsigned long,unsigned long,void*,unsigned long,unsigned long,void*)'
  extern 'void* CreateFileW(char*,unsigned long,unsigned long,void*,unsigned long,unsigned long,void*)'
  extern 'int CloseHandle(void*)'
  extern 'int GetFileInformationByHandle(void*,void*)'
  extern 'int FileTimeToSystemTime(void*,void*)'
  extern 'int GetFileTime(void*,void*,void*,void*)'

  INVALID_HANDLE_VALUE = 4294967295

  # Access rights constants
  FILE_READ_DATA = 1
  FILE_WRITE_DATA = 2
  FILE_APPEND_DATA = 4
  GENERIC_READ = 0x80000000
  GENERIC_WRITE = 0x40000000
  GENERIC_EXECUTE = 0x20000000
  GENERIC_ALL = 0x10000000

  # Share mode constants
  FILE_SHARE_0 = 0x00000000 # Prevents other processes from opening a file or device if they request delete, read, or write access.
  FILE_SHARE_DELETE = 0x00000004 # Enables subsequent open operations on a file or device to request delete access.
  FILE_SHARE_READ = 0x00000001 # Enables subsequent open operations on a file or device to request read access.
  FILE_SHARE_WRITE = 0x00000002 # Enables subsequent open operations on a file or device to request write access.

  # Creation disposition constants
  CREATE_ALWAYS = 2 # Creates a new file, always.
  CREATE_NEW = 1 # Creates a new file, only if it does not already exist.
  OPEN_ALWAYS = 4 # Opens a file, always.
  OPEN_EXISTING = 3 # Opens a file or device, only if it exists.
  TRUNCATE_EXISTING = 5 # Opens a file and truncates it so that its size is zero bytes, only if it exists

  # CreateFile flags and attributes constants
  FILE_ATTRIBUTE_READONLY = 1 # The file is read only. Applications can read the file, but cannot write to or delete it.
  FILE_ATTRIBUTE_HIDDEN = 2 # The file is hidden. Do not include it in an ordinary directory listing.
  FILE_ATTRIBUTE_SYSTEM = 4 # The file is part of or used exclusively by an operating system.
  FILE_ATTRIBUTE_DIRECTORY = 16 # The handle that identifies a directory.
  FILE_ATTRIBUTE_ARCHIVE = 32 # The file should be archived. Applications use this attribute to mark files for backup or removal.
  FILE_ATTRIBUTE_NORMAL = 128 # The file does not have other attributes set. This attribute is valid only if used alone.
  FILE_ATTRIBUTE_TEMPORARY = 256 # The file is being used for temporary storage.
  FILE_ATTRIBUTE_OFFLINE = 4096 # The data of a file is not immediately available.
  FILE_ATTRIBUTE_ENCRYPTED = 16384 # The file or directory is encrypted.

  FILE_FLAG_BACKUP_SEMANTICS = 0x02000000 # The file is being opened or created for a backup or restore operation.
  FILE_FLAG_DELETE_ON_CLOSE = 0x04000000 # The file is to be deleted immediately after all of its handles are closed, which includes the specified handle and any other open or duplicated handles.
  FILE_FLAG_NO_BUFFERING = 0x20000000 # The file or device is being opened with no system caching for data reads and writes.
  FILE_FLAG_OPEN_NO_RECALL = 0x00100000 # The file data is requested, but it should continue to be located in remote storage.
  FILE_FLAG_OPEN_REPARSE_POINT = 0x00200000 # Normal reparse point processing will not occur; CreateFile will attempt to open the reparse point.
  FILE_FLAG_OVERLAPPED = 0x40000000 # The file or device is being opened or created for asynchronous I/O.
  FILE_FLAG_POSIX_SEMANTICS = 0x0100000 # Access will occur according to POSIX rules.
  FILE_FLAG_RANDOM_ACCESS = 0x10000000 # Access is intended to be random.
  FILE_FLAG_SESSION_AWARE = 0x00800000 # The file or device is being opened with session awareness.
  FILE_FLAG_SEQUENTIAL_SCAN = 0x08000000 # Access is intended to be sequential from beginning to end.
  FILE_FLAG_WRITE_THROUGH = 0x80000000 # Write operations will not go through any intermediate cache, they will go directly to disk.

  # CreateFile (retrieve metadata only)
  def self.create_file_metadata(file_path)
    begin
      Kernel32Lib.CreateFile(file_path.encode('ascii-8bit'),Kernel32Lib::GENERIC_READ,Kernel32Lib::FILE_SHARE_READ | Kernel32Lib::FILE_SHARE_WRITE,nil,Kernel32Lib::OPEN_EXISTING,0,nil)
    rescue Encoding::UndefinedConversionError
      Kernel32Lib.CreateFileW(file_path.encode('utf-16le'),Kernel32Lib::GENERIC_READ,Kernel32Lib::FILE_SHARE_READ | Kernel32Lib::FILE_SHARE_WRITE,nil,Kernel32Lib::OPEN_EXISTING,0,nil)
    end
  end

  # FileTimeToSystemTime
  def self.file_time_to_system_time(file_time)
    lp_system_time = '0'*16
    Kernel32Lib.FileTimeToSystemTime(file_time,lp_system_time)
    lp_system_time = lp_system_time.unpack('S8')
    Time.mktime(lp_system_time[0],lp_system_time[1],lp_system_time[3],lp_system_time[4],lp_system_time[5],lp_system_time[6]+lp_system_time[7]/1000)
  end

  # GetFileInformationByHandle
  def self.get_file_information_by_handle(file_path)
    lp_file_information = '0'*52

    handle = create_file_metadata(file_path)
    raise Kernel32Lib::Error, Kernel32Lib.get_last_error_message if handle.to_i == Kernel32Lib::INVALID_HANDLE_VALUE

    result = Kernel32Lib.GetFileInformationByHandle(handle,lp_file_information)
    raise Kernel32Lib::Error, Kernel32Lib.get_last_error_message if result == 0

    Kernel32Lib.CloseHandle(handle)

    lp_file_information = lp_file_information.unpack('I13')
    {
      file_path: file_path,
      file_attributes: lp_file_information[0],
      creation_time: Kernel32Lib.file_time_to_system_time(lp_file_information[1,2].pack('I2')),
      last_access_time: Kernel32Lib.file_time_to_system_time(lp_file_information[3,4].pack('I2')),
      last_write_time: Kernel32Lib.file_time_to_system_time(lp_file_information[5,6].pack('I2')),
      volume_serial_number: lp_file_information[7],
      file_size_high: lp_file_information[8],
      file_size_low: lp_file_information[9],
      number_of_links: lp_file_information[10],
      file_index_high: lp_file_information[11],
      file_index_low: lp_file_information[12]
    }
  end

  # GetFileTime
  def self.get_file_time(file_path)
    lp_creation_time = '0'*8
    lp_access_time = '0'*8
    lp_write_time = '0'*8


    #handle = Kernel32Lib.CreateFile(file_path,Kernel32Lib::GENERIC_READ,Kernel32Lib::FILE_SHARE_READ | Kernel32Lib::FILE_SHARE_WRITE,nil,Kernel32Lib::OPEN_EXISTING,0,nil)
    handle = create_file_metadata(file_path)
    raise Kernel32Lib::Error, Kernel32Lib.get_last_error_message if handle.to_i == Kernel32Lib::INVALID_HANDLE_VALUE

    result = Kernel32Lib.GetFileTime(handle,lp_creation_time,lp_access_time,lp_write_time)
    raise Kernel32Lib::Error, Kernel32Lib.get_last_error_message if result == 0

    Kernel32Lib.CloseHandle(handle)

    lp_creation_time = lp_creation_time.unpack('I2')
    lp_access_time = lp_access_time.unpack('I2')
    lp_write_time = lp_write_time.unpack('I2')

    {
        file_path: file_path,
        creation_time: Kernel32Lib.file_time_to_system_time(lp_creation_time.pack('I2')),
        last_access_time: Kernel32Lib.file_time_to_system_time(lp_access_time.pack('I2')),
        last_write_time: Kernel32Lib.file_time_to_system_time(lp_write_time.pack('I2')),
    }
  end
end


lib/kernel32lib.rb
# encoding: utf-8

require 'fiddle'
require 'fiddle/import'

# Kernel32Lib module
# supports kernel32 WinAPI functions
module Kernel32Lib
  extend Fiddle::Importer
  dlload 'kernel32.dll'
  extern 'unsigned long GetLastError(void)'

  class Error < RuntimeError

  end

  ERROR_CODES = []
  ERROR_CODES[0]='The operation completed successfully.'
  ERROR_CODES[1]='Incorrect function.'
  ERROR_CODES[2]='The system cannot find the file specified.'
  ERROR_CODES[3]='The system cannot find the path specified.'
  ERROR_CODES[4]='The system cannot open the file.'
  ERROR_CODES[5]='Access is denied.'
  ERROR_CODES[6]='The handle is invalid.'
  ERROR_CODES[7]='The storage control blocks were destroyed.'
  ERROR_CODES[8]='Not enough storage is available to process this command.'
  ERROR_CODES[9]='The storage control block address is invalid.'
  ERROR_CODES[10]='The environment is incorrect.'
  ERROR_CODES[11]='An attempt was made to load a program with an incorrect format.'
  ERROR_CODES[12]='The access code is invalid.'
  ERROR_CODES[13]='The data is invalid.'
  ERROR_CODES[14]='Not enough storage is available to complete this operation.'
  ERROR_CODES[15]='The system cannot find the drive specified.'
  ERROR_CODES[16]='The directory cannot be removed.'
  ERROR_CODES[17]='The system cannot move the file to a different disk drive.'
  ERROR_CODES[18]='There are no more files.'
  ERROR_CODES[19]='The media is write protected.'
  ERROR_CODES[20]='The system cannot find the device specified.'
  ERROR_CODES[21]='The device is not ready.'
  ERROR_CODES[22]='The device does not recognize the command.'
  ERROR_CODES[23]='Data error (cyclic redundancy check).'
  ERROR_CODES[24]='The program issued a command but the command length is incorrect.'
  ERROR_CODES[25]='The drive cannot locate a specific area or track on the disk.'
  ERROR_CODES[26]='The specified disk or diskette cannot be accessed.'
  ERROR_CODES[27]='The drive cannot find the sector requested.'
  ERROR_CODES[28]='The printer is out of paper.'
  ERROR_CODES[29]='The system cannot write to the specified device.'
  ERROR_CODES[30]='The system cannot read from the specified device.'
  ERROR_CODES[31]='A device attached to the system is not functioning.'
  ERROR_CODES[32]='The process cannot access the file because it is being used by another process.'
  ERROR_CODES[33]='The process cannot access the file because another process has locked a portion of the file.'
  ERROR_CODES[34]='The wrong diskette is in the drive. Insert %2 (Volume Serial Number: %3) into drive %1.'
  ERROR_CODES[36]='Too many files opened for sharing.'
  ERROR_CODES[38]='Reached the end of the file.'
  ERROR_CODES[39]='The disk is full.'
  ERROR_CODES[50]='The request is not supported.'
  ERROR_CODES[51]='Windows cannot find the network path. Verify that the network path is correct and the destination computer is not busy or turned off. If Windows still cannot find the network path, contact your network administrator.'
  ERROR_CODES[52]='You were not connected because a duplicate name exists on the network. If joining a domain, go to System in Control Panel to change the computer name and try again. If joining a workgroup, choose another workgroup name.'
  ERROR_CODES[53]='The network path was not found.'
  ERROR_CODES[54]='The network is busy.'
  ERROR_CODES[55]='The specified network resource or device is no longer available.'
  ERROR_CODES[56]='The network BIOS command limit has been reached.'
  ERROR_CODES[57]='A network adapter hardware error occurred.'
  ERROR_CODES[58]='The specified server cannot perform the requested operation.'
  ERROR_CODES[59]='An unexpected network error occurred.'
  ERROR_CODES[60]='The remote adapter is not compatible.'
  ERROR_CODES[61]='The printer queue is full.'
  ERROR_CODES[62]='Space to store the file waiting to be printed is not available on the server.'
  ERROR_CODES[63]='Your file waiting to be printed was deleted.'
  ERROR_CODES[64]='The specified network name is no longer available.'
  ERROR_CODES[65]='Network access is denied.'
  ERROR_CODES[66]='The network resource type is not correct.'
  ERROR_CODES[67]='The network name cannot be found.'
  ERROR_CODES[68]='The name limit for the local computer network adapter card was exceeded.'
  ERROR_CODES[69]='The network BIOS session limit was exceeded.'
  ERROR_CODES[70]='The remote server has been paused or is in the process of being started.'
  ERROR_CODES[71]='No more connections can be made to this remote computer at this time because there are already as many connections as the computer can accept.'
  ERROR_CODES[72]='The specified printer or disk device has been paused.'
  ERROR_CODES[80]='The file exists.'
  ERROR_CODES[82]='The directory or file cannot be created.'
  ERROR_CODES[83]='Fail on INT 24.'
  ERROR_CODES[84]='Storage to process this request is not available.'
  ERROR_CODES[85]='The local device name is already in use.'
  ERROR_CODES[86]='The specified network password is not correct.'
  ERROR_CODES[87]='The parameter is incorrect.'
  ERROR_CODES[88]='A write fault occurred on the network.'
  ERROR_CODES[89]='The system cannot start another process at this time.'
  ERROR_CODES[100]='Cannot create another system semaphore.'
  ERROR_CODES[101]='The exclusive semaphore is owned by another process.'
  ERROR_CODES[102]='The semaphore is set and cannot be closed.'
  ERROR_CODES[103]='The semaphore cannot be set again.'
  ERROR_CODES[104]='Cannot request exclusive semaphores at interrupt time.'
  ERROR_CODES[105]='The previous ownership of this semaphore has ended.'
  ERROR_CODES[106]='Insert the diskette for drive %1.'
  ERROR_CODES[107]='The program stopped because an alternate diskette was not inserted.'
  ERROR_CODES[108]='The disk is in use or locked by another process.'
  ERROR_CODES[109]='The pipe has been ended.'
  ERROR_CODES[110]='The system cannot open the device or file specified.'
  ERROR_CODES[111]='The file name is too long.'
  ERROR_CODES[112]='There is not enough space on the disk.'
  ERROR_CODES[113]='No more internal file identifiers available.'
  ERROR_CODES[114]='The target internal file identifier is incorrect.'
  ERROR_CODES[117]='The IOCTL call made by the application program is not correct.'
  ERROR_CODES[118]='The verify-on-write switch parameter value is not correct.'
  ERROR_CODES[119]='The system does not support the command requested.'
  ERROR_CODES[120]='This function is not supported on this system.'
  ERROR_CODES[121]='The semaphore timeout period has expired.'
  ERROR_CODES[122]='The data area passed to a system call is too small.'
  ERROR_CODES[123]='The filename, directory name, or volume label syntax is incorrect.'
  ERROR_CODES[124]='The system call level is not correct.'
  ERROR_CODES[125]='The disk has no volume label.'
  ERROR_CODES[126]='The specified module could not be found.'
  ERROR_CODES[127]='The specified procedure could not be found.'
  ERROR_CODES[128]='There are no child processes to wait for.'
  ERROR_CODES[129]='The %1 application cannot be run in Win32 mode.'
  ERROR_CODES[130]='Attempt to use a file handle to an open disk partition for an operation other than raw disk I/O.'
  ERROR_CODES[131]='An attempt was made to move the file pointer before the beginning of the file.'
  ERROR_CODES[132]='The file pointer cannot be set on the specified device or file.'
  ERROR_CODES[133]='A JOIN or SUBST command cannot be used for a drive that contains previously joined drives.'
  ERROR_CODES[134]='An attempt was made to use a JOIN or SUBST command on a drive that has already been joined.'
  ERROR_CODES[135]='An attempt was made to use a JOIN or SUBST command on a drive that has already been substituted.'
  ERROR_CODES[136]='The system tried to delete the JOIN of a drive that is not joined.'
  ERROR_CODES[137]='The system tried to delete the substitution of a drive that is not substituted.'
  ERROR_CODES[138]='The system tried to join a drive to a directory on a joined drive.'
  ERROR_CODES[139]='The system tried to substitute a drive to a directory on a substituted drive.'
  ERROR_CODES[140]='The system tried to join a drive to a directory on a substituted drive.'
  ERROR_CODES[141]='The system tried to SUBST a drive to a directory on a joined drive.'
  ERROR_CODES[142]='The system cannot perform a JOIN or SUBST at this time.'
  ERROR_CODES[143]='The system cannot join or substitute a drive to or for a directory on the same drive.'
  ERROR_CODES[144]='The directory is not a subdirectory of the root directory.'
  ERROR_CODES[145]='The directory is not empty.'
  ERROR_CODES[146]='The path specified is being used in a substitute.'
  ERROR_CODES[147]='Not enough resources are available to process this command.'
  ERROR_CODES[148]='The path specified cannot be used at this time.'
  ERROR_CODES[149]='An attempt was made to join or substitute a drive for which a directory on the drive is the target of a previous substitute.'
  ERROR_CODES[150]='System trace information was not specified in your CONFIG.SYS file, or tracing is disallowed.'
  ERROR_CODES[151]='The number of specified semaphore events for DosMuxSemWait is not correct.'
  ERROR_CODES[152]='DosMuxSemWait did not execute; too many semaphores are already set.'
  ERROR_CODES[153]='The DosMuxSemWait list is not correct.'
  ERROR_CODES[154]='The volume label you entered exceeds the label character limit of the target file system.'
  ERROR_CODES[155]='Cannot create another thread.'
  ERROR_CODES[156]='The recipient process has refused the signal.'
  ERROR_CODES[157]='The segment is already discarded and cannot be locked.'
  ERROR_CODES[158]='The segment is already unlocked.'
  ERROR_CODES[159]='The address for the thread ID is not correct.'
  ERROR_CODES[160]='One or more arguments are not correct.'
  ERROR_CODES[161]='The specified path is invalid.'
  ERROR_CODES[162]='A signal is already pending.'
  ERROR_CODES[164]='No more threads can be created in the system.'
  ERROR_CODES[167]='Unable to lock a region of a file.'
  ERROR_CODES[170]='The requested resource is in use.'
  ERROR_CODES[171]="Device's command support detection is in progress."
  ERROR_CODES[173]='A lock request was not outstanding for the supplied cancel region.'
  ERROR_CODES[174]='The file system does not support atomic changes to the lock type.'
  ERROR_CODES[180]='The system detected a segment number that was not correct.'
  ERROR_CODES[182]='The operating system cannot run %1.'
  ERROR_CODES[183]='Cannot create a file when that file already exists.'
  ERROR_CODES[186]='The flag passed is not correct.'
  ERROR_CODES[187]='The specified system semaphore name was not found.'
  ERROR_CODES[188]='The operating system cannot run %1.'
  ERROR_CODES[189]='The operating system cannot run %1.'
  ERROR_CODES[190]='The operating system cannot run %1.'
  ERROR_CODES[191]='Cannot run %1 in Win32 mode.'
  ERROR_CODES[192]='The operating system cannot run %1.'
  ERROR_CODES[193]='%1 is not a valid Win32 application.'
  ERROR_CODES[194]='The operating system cannot run %1.'
  ERROR_CODES[195]='The operating system cannot run %1.'
  ERROR_CODES[196]='The operating system cannot run this application program.'
  ERROR_CODES[197]='The operating system is not presently configured to run this application.'
  ERROR_CODES[198]='The operating system cannot run %1.'
  ERROR_CODES[199]='The operating system cannot run this application program.'
  ERROR_CODES[200]='The code segment cannot be greater than or equal to 64K.'
  ERROR_CODES[201]='The operating system cannot run %1.'
  ERROR_CODES[202]='The operating system cannot run %1.'
  ERROR_CODES[203]='The system could not find the environment option that was entered.'
  ERROR_CODES[205]='No process in the command subtree has a signal handler.'
  ERROR_CODES[206]='The filename or extension is too long.'
  ERROR_CODES[207]='The ring 2 stack is in use.'
  ERROR_CODES[208]='The global filename characters, * or ?, are entered incorrectly or too many global filename characters are specified.'
  ERROR_CODES[209]='The signal being posted is not correct.'
  ERROR_CODES[210]='The signal handler cannot be set.'
  ERROR_CODES[212]='The segment is locked and cannot be reallocated.'
  ERROR_CODES[214]='Too many dynamic-link modules are attached to this program or dynamic-link module.'
  ERROR_CODES[215]='Cannot nest calls to LoadModule.'
  ERROR_CODES[216]="This version of %1 is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher."
  ERROR_CODES[217]='The image file %1 is signed, unable to modify.'
  ERROR_CODES[218]='The image file %1 is strong signed, unable to modify.'
  ERROR_CODES[220]='This file is checked out or locked for editing by another user.'
  ERROR_CODES[221]='The file must be checked out before saving changes.'
  ERROR_CODES[222]='The file type being saved or retrieved has been blocked.'
  ERROR_CODES[223]='The file size exceeds the limit allowed and cannot be saved.'
  ERROR_CODES[224]='Access Denied. Before opening files in this location, you must first add the web site to your trusted sites list, browse to the web site, and select the option to login automatically.'
  ERROR_CODES[225]='Operation did not complete successfully because the file contains a virus or potentially unwanted software.'
  ERROR_CODES[226]='This file contains a virus or potentially unwanted software and cannot be opened. Due to the nature of this virus or potentially unwanted software, the file has been removed from this location.'
  ERROR_CODES[229]='The pipe is local.'
  ERROR_CODES[230]='The pipe state is invalid.'
  ERROR_CODES[231]='All pipe instances are busy.'
  ERROR_CODES[232]='The pipe is being closed.'
  ERROR_CODES[233]='No process is on the other end of the pipe.'
  ERROR_CODES[234]='More data is available.'
  ERROR_CODES[240]='The session was canceled.'
  ERROR_CODES[254]='The specified extended attribute name was invalid.'
  ERROR_CODES[255]='The extended attributes are inconsistent.'
  ERROR_CODES[258]='The wait operation timed out.'
  ERROR_CODES[259]='No more data is available.'
  ERROR_CODES[266]='The copy functions cannot be used.'
  ERROR_CODES[267]='The directory name is invalid.'
  ERROR_CODES[275]='The extended attributes did not fit in the buffer.'
  ERROR_CODES[276]='The extended attribute file on the mounted file system is corrupt.'
  ERROR_CODES[277]='The extended attribute table file is full.'
  ERROR_CODES[278]='The specified extended attribute handle is invalid.'
  ERROR_CODES[282]='The mounted file system does not support extended attributes.'
  ERROR_CODES[288]='Attempt to release mutex not owned by caller.'
  ERROR_CODES[298]='Too many posts were made to a semaphore.'
  ERROR_CODES[299]='Only part of a ReadProcessMemory or WriteProcessMemory request was completed.'
  ERROR_CODES[300]='The oplock request is denied.'
  ERROR_CODES[301]='An invalid oplock acknowledgment was received by the system.'
  ERROR_CODES[302]='The volume is too fragmented to complete this operation.'
  ERROR_CODES[303]='The file cannot be opened because it is in the process of being deleted.'
  ERROR_CODES[304]='Short name settings may not be changed on this volume due to the global registry setting.'
  ERROR_CODES[305]='Short names are not enabled on this volume.'
  ERROR_CODES[306]='The security stream for the given volume is in an inconsistent state. Please run CHKDSK on the volume.'
  ERROR_CODES[307]='A requested file lock operation cannot be processed due to an invalid byte range.'
  ERROR_CODES[308]='The subsystem needed to support the image type is not present.'
  ERROR_CODES[309]='The specified file already has a notification GUID associated with it.'
  ERROR_CODES[310]='An invalid exception handler routine has been detected.'
  ERROR_CODES[311]='Duplicate privileges were specified for the token.'
  ERROR_CODES[312]='No ranges for the specified operation were able to be processed.'
  ERROR_CODES[313]='Operation is not allowed on a file system internal file.'
  ERROR_CODES[314]='The physical resources of this disk have been exhausted.'
  ERROR_CODES[315]='The token representing the data is invalid.'
  ERROR_CODES[316]='The device does not support the command feature.'
  ERROR_CODES[317]='The system cannot find message text for message number 0x%1 in the message file for %2.'
  ERROR_CODES[318]='The scope specified was not found.'
  ERROR_CODES[319]='The Central Access Policy specified is not defined on the target machine.'
  ERROR_CODES[320]='The Central Access Policy obtained from Active Directory is invalid.'
  ERROR_CODES[321]='The device is unreachable.'
  ERROR_CODES[322]='The target device has insufficient resources to complete the operation.'
  ERROR_CODES[323]='A data integrity checksum error occurred. Data in the file stream is corrupt.'
  ERROR_CODES[324]='An attempt was made to modify both a KERNEL and normal Extended Attribute (EA) in the same operation.'
  ERROR_CODES[326]='Device does not support file-level TRIM.'
  ERROR_CODES[327]="The command specified a data offset that does not align to the device's granularity/alignment."
  ERROR_CODES[328]='The command specified an invalid field in its parameter list.'
  ERROR_CODES[329]='An operation is currently in progress with the device.'
  ERROR_CODES[330]='An attempt was made to send down the command via an invalid path to the target device.'
  ERROR_CODES[331]='The command specified a number of descriptors that exceeded the maximum supported by the device.'
  ERROR_CODES[332]='Scrub is disabled on the specified file.'
  ERROR_CODES[333]='The storage device does not provide redundancy.'
  ERROR_CODES[334]='An operation is not supported on a resident file.'
  ERROR_CODES[335]='An operation is not supported on a compressed file.'
  ERROR_CODES[336]='An operation is not supported on a directory.'
  ERROR_CODES[337]='The specified copy of the requested data could not be read.'
  ERROR_CODES[350]='No action was taken as a system reboot is required.'
  ERROR_CODES[351]='The shutdown operation failed.'
  ERROR_CODES[352]='The restart operation failed.'
  ERROR_CODES[353]='The maximum number of sessions has been reached.'
  ERROR_CODES[400]='The thread is already in background processing mode.'
  ERROR_CODES[401]='The thread is not in background processing mode.'
  ERROR_CODES[402]='The process is already in background processing mode.'
  ERROR_CODES[403]='The process is not in background processing mode.'
  ERROR_CODES[487]='Attempt to access invalid address.'

  # GetLastError
  def self.get_last_error
    Kernel32Lib.GetLastError(0)
  end

  def self.translate_error(error_code)
    (ERROR_CODES[error_code] || "Error message by code #{error_code} not found.")
  end

  def self.get_last_error_message
    translate_error(get_last_error)
  end
end


win_api/consoles.rb
# encoding: utf-8

require(File.expand_path('../../lib/kernel32lib.rb',__FILE__))

# Kernel32Lib module
# Consoles
module Kernel32Lib
  extern 'void* GetStdHandle(unsigned long)'
  extern 'int SetConsoleCursorPosition(void*,void*)'
  extern 'int GetConsoleScreenBufferInfo(void*,void*)'
  extern 'int SetConsoleScreenBufferSize(void*,void*)'
  extern 'int SetConsoleWindowInfo(void*,void,void*)'
  extern 'int SetConsoleTextAttribute(void*,unsigned long)'
  extern 'unsigned int GetConsoleOutputCP(void)'
  extern 'int SetConsoleOutputCP(unsigned int)'

  STD_INPUT_HANDLE = -10
  STD_OUTPUT_HANDLE = -11
  STD_ERROR_HANDLE = -12

  # Attributes flags:
  FOREGROUND_BLUE = 0x0001 # text color contains blue.
  FOREGROUND_GREEN = 0x0002 # text color contains green.
  FOREGROUND_RED = 0x0004 # text color contains red.
  FOREGROUND_WHITE = 0x0007 # text color contains white on black background.
  FOREGROUND_INTENSITY = 0x0008 # text color is intensified.
  BACKGROUND_BLUE = 0x0010 # background color contains blue.
  BACKGROUND_GREEN = 0x0020 # background color contains green.
  BACKGROUND_RED = 0x0040 # background color contains red.
  BACKGROUND_WHITE = 0x0070 # background color contains white with black foreground.
  BACKGROUND_INTENSITY = 0x0080 # background color is intensified.
  COMMON_LVB_LEADING_BYTE = 0x0100 # Leading Byte of DBCS
  COMMON_LVB_TRAILING_BYTE = 0x0200 # Trailing Byte of DBCS
  COMMON_LVB_GRID_HORIZONTAL = 0x0400 # DBCS: Grid attribute: top horizontal.
  COMMON_LVB_GRID_LVERTICAL = 0x0800 # DBCS: Grid attribute: left vertical.
  COMMON_LVB_GRID_RVERTICAL = 0x1000 # DBCS: Grid attribute: right vertical.
  COMMON_LVB_REVERSE_VIDEO = 0x4000 # DBCS: Reverse fore/back ground attribute.
  COMMON_LVB_UNDERSCORE = 0x8000 # DBCS: Underscore.
  COMMON_LVB_SBCSDBCS = 0x0300 # SBCS or DBCS flag.

  # SetConsoleCursorPosition
  def self.set_console_cursor_position(console_output,cursor_position)
    lp_cursor_position = cursor_position.pack('s2')
    Kernel32Lib.SetConsoleCursorPosition(console_output,lp_cursor_position.unpack('i')[0].to_i)
  end

  # GetConsoleScreenBufferInfo
  def self.get_console_screen_buffer_info(console_output)
    lp_console_screen_buffer_info = '0'*22
    Kernel32Lib.GetConsoleScreenBufferInfo(console_output,lp_console_screen_buffer_info)
    lp_console_screen_buffer_info = lp_console_screen_buffer_info.unpack('s4Ss4s2')
    console_screen_buffer_info = {
        size: [ lp_console_screen_buffer_info[0].to_i, lp_console_screen_buffer_info[1].to_i ],
        cursor_position: [ lp_console_screen_buffer_info[2], lp_console_screen_buffer_info[3].to_i ],
        attributes: lp_console_screen_buffer_info[4].to_i,
        window: [ lp_console_screen_buffer_info[5].to_i, lp_console_screen_buffer_info[6].to_i, lp_console_screen_buffer_info[7].to_i, lp_console_screen_buffer_info[8].to_i ],
        maximum_window_size: [ lp_console_screen_buffer_info[9].to_i, lp_console_screen_buffer_info[10].to_i ]
    }
    console_screen_buffer_info
  end

  # SetConsoleTextAttribute
  def self.set_console_text_attribute(console_output,attributes)
    Kernel32Lib.SetConsoleTextAttribute(console_output,attributes)
  end

  # GetStdHandle
  def self.get_std_handle(std_handle)
    Kernel32Lib.GetStdHandle(std_handle)
  end

  # GetConsoleOutputCP
  def self.get_console_output_cp
    Kernel32Lib.GetConsoleOutputCP(0)
  end

  # SetConsoleOutputCP
  def self.set_console_output_cp(code_page_id)
    Kernel32Lib.SetConsoleOutputCP(code_page_id)
  end

  # SetConsoleScreenBufferSize
  def self.set_console_screen_buffer_size(console_output,size)
    lp_size = size.pack('s2')
    Kernel32Lib.SetConsoleScreenBufferSize(console_output,lp_size.unpack('i')[0].to_i)
  end

  # SetConsoleWindowInfo
  def self.set_console_window_info(console_output,absolute,console_window)
    lp_console_window = console_window.pack('s4')
    Kernel32Lib.SetConsoleWindowInfo(console_output,absolute,lp_console_window)
  end

end

fixtures/fixture_kernel32lib.rb
# encoding: utf-8

#require 'kernel32lib'
require(File.expand_path('../../win_api/files',__FILE__))
require(File.expand_path('../../win_api/consoles',__FILE__))

# Fixture module
# test fixture
module Fixture
  def self.show
    system 'cls'
    puts '# get_file_information_by_handle'
    puts "  - file information: #{Kernel32Lib.get_file_information_by_handle(File.join(File.expand_path('../fixtures'),'metadata_target_file.txt'))}"
    puts
    puts '# get_file_time'
    puts "  - file time: #{Kernel32Lib.get_file_time(File.join(File.expand_path('../fixtures'),'metadata_target_file.txt'))}"
    puts
    puts '# get_file_information_by_handle'
    puts "  - file information: #{Kernel32Lib.get_file_information_by_handle(File.join(File.expand_path('../fixtures'),'целевой_файл_с_метаданными.txt'))}"
    puts
    puts '# get_file_time'
    puts "  - file time: #{Kernel32Lib.get_file_time(File.join(File.expand_path('../fixtures'),'целевой_файл_с_метаданными.txt'))}"
    puts
    puts '# try console'
    console_output = Kernel32Lib.GetStdHandle(Kernel32Lib::STD_OUTPUT_HANDLE)
    init_console_screen_buffer_info = Kernel32Lib.get_console_screen_buffer_info(console_output)
    init_attributes = init_console_screen_buffer_info[:attributes]
    init_screen_buffer_size = init_console_screen_buffer_info[:size]
    init_window = init_console_screen_buffer_info[:window]
    puts "  - console screen buffer info (1): #{init_console_screen_buffer_info}"
    puts "  - console output code page: #{Kernel32Lib.get_console_output_cp}"
    Kernel32Lib.set_console_window_info(console_output,true,[0,0,1,1])
    Kernel32Lib.set_console_screen_buffer_size(console_output,[120,80])
    Kernel32Lib.set_console_window_info(console_output,true,[0,0,70,25])
    Kernel32Lib.set_console_text_attribute(console_output,Kernel32Lib::FOREGROUND_GREEN | Kernel32Lib::FOREGROUND_INTENSITY | Kernel32Lib::BACKGROUND_BLUE)
    puts "  - console screen buffer info (2): #{Kernel32Lib.get_console_screen_buffer_info(console_output)}"
    9.times { print '.'; sleep(0.5) }
    Kernel32Lib.set_console_text_attribute(console_output,init_attributes.to_i)
    Kernel32Lib.set_console_window_info(console_output,true,[0,0,1,1])
    Kernel32Lib.set_console_screen_buffer_size(console_output,init_screen_buffer_size)
    Kernel32Lib.set_console_window_info(console_output,true,init_window)
    puts
  end

end